### ISSUE TITLE
Expected CNN input, got InputTypeFeedForward

### ISSUE THREAD

**Issue by xiptos** [deeplearning4j/deeplearning4j#6245]
When importing a Keras model, I believe that DL4J is mistakingly adding a FF layer instead of a CNN, according to the following exception:

[main] INFO org.deeplearning4j.nn.modelimport.keras.Hdf5Archive - Unexpected end-of-input: was expecting closing '"' for name
 at [Source: {"class_name": "Model", "config": {"layers": [{"class_name": "InputLayer", "config": {"dtype": "float32", "batch_input_shape": [null, null, null, 3], "name": "input_3", "sparse": false}, "inbound_nodes": [], "name": "input_3"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv1", "bias_regularizer": null, "bias_constraint": null, "activation": "linear", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 10, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [3, 3]}, "inbound_nodes": [[["input_3", 0, 0, {}]]], "name": "conv1"}, {"class_name": "PReLU", "config": {"name": "PReLU1", "alpha_constraint": null, "trainable": true, "alpha_regularizer": null, "alpha_initializer": {"class_name": "Zeros", "config": {}}, "shared_axes": [1, 2]}, "inbound_nodes": [[["conv1", 0, 0, {}]]], "name": "PReLU1"}, {"class_name": "MaxPooling2D", "config": {"name": "max_pooling2d_6", "trainable": true, "data_format": "channels_last", "pool_size": [2, 2], "padding": "valid", "strides": [2, 2]}, "inbound_nodes": [[["PReLU1", 0, 0, {}]]], "name": "max_pooling2d_6"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv2", "bias_regularizer": null, "bias_constraint": null, "activation": "linear", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 16, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [3, 3]}, "inbound_nodes; line: 1, column: 4001]
[main] INFO org.deeplearning4j.nn.modelimport.keras.Hdf5Archive - Unexpected end-of-input: was expecting closing quote for a string value
 at [Source: {"class_name": "Model", "config": {"layers": [{"class_name": "InputLayer", "config": {"dtype": "float32", "batch_input_shape": [null, null, null, 3], "name": "input_3", "sparse": false}, "inbound_nodes": [], "name": "input_3"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv1", "bias_regularizer": null, "bias_constraint": null, "activation": "linear", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 10, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [3, 3]}, "inbound_nodes": [[["input_3", 0, 0, {}]]], "name": "conv1"}, {"class_name": "PReLU", "config": {"name": "PReLU1", "alpha_constraint": null, "trainable": true, "alpha_regularizer": null, "alpha_initializer": {"class_name": "Zeros", "config": {}}, "shared_axes": [1, 2]}, "inbound_nodes": [[["conv1", 0, 0, {}]]], "name": "PReLU1"}, {"class_name": "MaxPooling2D", "config": {"name": "max_pooling2d_6", "trainable": true, "data_format": "channels_last", "pool_size": [2, 2], "padding": "valid", "strides": [2, 2]}, "inbound_nodes": [[["PReLU1", 0, 0, {}]]], "name": "max_pooling2d_6"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv2", "bias_regularizer": null, "bias_constraint": null, "activation": "linear", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 16, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [3, 3]}, "inbound_nodes": [[["max_pooling2d_6", 0, 0, {}]]], "name": "conv2"}, {"class_name": "PReLU", "config": {"name": "PReLU2", "alpha_constraint": null, "trainable": true, "alpha_regularizer": null, "alpha_initializer": {"class_name": "Zeros", "config": {}}, "shared_axes": [1, 2]}, "inbound_nodes": [[["conv2", 0, 0, {}]]], "name": "PReLU2"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv3", "bias_regularizer": null, "bias_constraint": null, "activation": "linear", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 32, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [3, 3]}, "inbound_nodes": [[["PReLU2", 0, 0, {}]]], "name": "conv3"}, {"class_name": "PReLU", "config": {"name": "PReLU3", "alpha_constraint": null, "trainable": true, "alpha_regularizer": null, "alpha_initializer": {"class_name": "Zeros", "config": {}}, "shared_axes": [1, 2]}, "inbound_nodes": [[["conv3", 0, 0, {}]]], "name": "PReLU3"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"distribution": "uniform", "scale": 1.0, "seed": null, "mode": "fan_avg"}}, "name": "conv4-1", "bias_regularizer": null, "bias_constraint": null, "activation": "softmax", "trainable": true, "data_format": "channels_last", "padding": "valid", "strides": [1, 1], "dilation_rate": [1, 1], "kernel_regularizer": null, "filters": 2, "bias_initializer": {"class_name": "Zeros", "config": {}}, "use_bias": true, "activity_regularizer": null, "kernel_size": [1, 1]}, "inbound_nodes": [[["PReLU3", 0, 0, {}]]], "name": "conv4-1"}, {"class_name": "Conv2D", "config": {"kernel_constraint": null, "kernel_initializer": {"class_name": "Var; line: 1, column: 8001]
[main] INFO org.nd4j.linalg.factory.Nd4jBackend - Loaded [CpuBackend] backend
[main] INFO org.nd4j.nativeblas.NativeOpsHolder - Number of threads used for NativeOps: 2
[main] INFO org.nd4j.nativeblas.Nd4jBlas - Number of threads used for BLAS: 2
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Backend used: [CPU]; OS: [Mac OS X]
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Cores: [4]; Memory: [1.8GB];
[main] INFO org.nd4j.linalg.api.ops.executioner.DefaultOpExecutioner - Blas vendor: [MKL]
[main] INFO org.deeplearning4j.nn.conf.layers.InputTypeUtil - Automatic addition of FF -> CNN preprocessors: not yet implemented (layer name: "max_pooling2d_6")
Exception in thread "main" java.lang.IllegalStateException: Invalid input for Subsampling layer (layer name="max_pooling2d_6"): Expected CNN input, got InputTypeFeedForward(1000)
	at org.deeplearning4j.nn.conf.layers.SubsamplingLayer.getOutputType(SubsamplingLayer.java:133)
	at org.deeplearning4j.nn.conf.graph.LayerVertex.getOutputType(LayerVertex.java:131)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration.getLayerActivationTypes(ComputationGraphConfiguration.java:455)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration.addPreProcessors(ComputationGraphConfiguration.java:368)
	at org.deeplearning4j.nn.conf.ComputationGraphConfiguration$GraphBuilder.build(ComputationGraphConfiguration.java:1040)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraphConfiguration(KerasModel.java:396)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:417)
	at org.deeplearning4j.nn.modelimport.keras.KerasModel.getComputationGraph(KerasModel.java:406)
	at pt.ipb.face.Mtcnn2Dl4j.importFunctionalModelH5Test(Mtcnn2Dl4j.java:45)
	at pt.ipb.face.Mtcnn2Dl4j.main(Mtcnn2Dl4j.java:26)

I'm using macOS High Sierra, with java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)

The models are in attach as the source code.

[Archive.zip](https://github.com/deeplearning4j/deeplearning4j/files/2309434/Archive.zip)
[Mtcnn2Dl4j.java.zip](https://github.com/deeplearning4j/deeplearning4j/files/2309437/Mtcnn2Dl4j.java.zip)

**Comment by maxpumperla** [comment]
@xiptos are you working on snapshots? this is precisely what I fixed in your last issue. If it doesn't work on snapshots, feel free to reopen.

The reason is that PReLU was a FeedForward layer before my patch, so DL4J adds a CNN to FF preprocessor automatically, hence the output of PReLU doesn't connect to the pooling layer.
